### PR TITLE
feat: add fn parse_checksummed

### PR DESCRIPTION
**Pull request by izayl**
## Motivation

as ethers-rs provide `to_checksum` function, but there not provide a way to verify a checksum-like address.
so I add a utilities function to verify a checksum address is valid.

## PR Checklist

-   [x] Added Tests
-   [ ] Added Documentation
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-core/src/utils/mod.rs
Comment (mattsse) [inline ethers-core/src/utils/mod.rs:442]: this should be false, no?

File: ethers-core/src/utils/mod.rs
Comment (mattsse) [inline ethers-core/src/utils/mod.rs:435]:
```suggestion
    if addr.chars().any(|c| c.is_ascii_uppercase()) {
```

File: ethers-core/src/utils/mod.rs
Comment (mattsse) [inline ethers-core/src/utils/mod.rs:424]: needs info when this returns true or false

**Review by mattsse — CHANGES_REQUESTED** [review]

File: ethers-core/src/utils/mod.rs
Comment (mattsse) [inline ethers-core/src/utils/mod.rs:425]: don't love the name,
also not a fan of the `to_checksum` function, should have been named `to_checksum_address`
but, to keep it consistent, this is fine I guess.

Comment (izayl) [reply ethers-core/src/utils/mod.rs:425]: maybe `check_address_checksum`?
ref: https://github.com/web3/web3.js/blob/aaf26c8806bc9fb60cf6dcb6658104963c6c7fc7/packages/web3-utils/src/Utils.js#L140

File: ethers-core/src/utils/mod.rs
Comment (izayl) [reply ethers-core/src/utils/mod.rs:442]: maybe the function name cause this confuse.
if the address is checksummed address, it need to do validation.
but if the address is not checksummed, it should not to do validation, so I chose to return `true`

File: ethers-core/src/utils/mod.rs
Comment (mattsse) [inline ethers-core/src/utils/mod.rs:438]: use 
```
if let Ok() {
let checksum_addr = to_checksum(&address.unwrap(), chain_id);

    return checksum_addr.strip_prefix("0x").unwrap_or(&checksum_addr) == addr

}
false
```

**Review by mattsse — CHANGES_REQUESTED** [review]
smol style nit

File: ethers-core/src/utils/mod.rs
Comment (izayl) [reply ethers-core/src/utils/mod.rs]: Done

File: ethers-core/src/utils/mod.rs
Comment (izayl) [reply ethers-core/src/utils/mod.rs:427]: like this?

```rs
/// Returns a `Result<Address>`: `Ok(Address)` if addr is checksummed, `Err()` otherwise
pub fn parse_checksummed(addr: &str, chain_id: Option<u8>) -> Result<Address>
```

Comment (prestwich) [reply ethers-core/src/utils/mod.rs:427]: function signature is awesome 👌 

I'd write docs similar to this:
```/// Parses an [EIP-1191](https://eips.ethereum.org/EIPS/eip-1191) checksum address. Returns `Ok(address)` if the checksummed address is valid, `Err()` otherwise. If `chain_id` is `None`, falls back to [EIP-55](https://eips.ethereum.org/EIPS/eip-55) address checksum method```

Comment (izayl) [reply ethers-core/src/utils/mod.rs]: @prestwich refactored, pls check it

File: ethers-core/src/utils/mod.rs
Comment (prestwich) [inline ethers-core/src/utils/mod.rs:441]: rather than returning text too long, I think we can do the following

1. add a hex error to the ConversionError enum
```
enum ConversionError { 
    ...,
    /// Hex error, usually from address parsing
    #[error(transparent)]
    FromHexError(#[from]hex::FromHexError)
}
```

2. change this function to use a `?`

```
        let addr: Address = addr.strip_prefix("0x").unwrap_or(addr).parse()?
        if checksum_addr.strip_prefix("0x").unwrap_or(&checksum_addr) == addr {
            return Ok(address)
        } else {
            return Err(ConversionError::InvalidAddressChecksum)
        }


```

**Comment by prestwich** [comment]
changes are good. 1 more small ergonomics tweak 🙏

File: ethers-core/src/utils/mod.rs
Comment (izayl) [reply ethers-core/src/utils/mod.rs:441]: the `FromHexError` is from `rustc_hex` crate, should I add a it?

Comment (izayl) [reply ethers-core/src/utils/mod.rs:441]: https://github.com/paritytech/parity-common/blob/master/fixed-hash/src/hash.rs#L568-L593

Comment (prestwich) [reply ethers-core/src/utils/mod.rs:441]: leave it to parity to make our lives more complicated 🤣 🙄

I think you should be able to make it work as follows:
1. reference the fully qualified associated type `<Address as std::str::FromStr>::Err` in the enum variant
2. instead of doing a `#[from]` in the error variant, use `map_err` to convert

```
// variant specifies that it takes "whatever error the `parse()` implementation says is the error"
FromHexError(<Address as std::str::FromStr>::Err)

// map err on the first line of the function
let addr: Address = addr.strip_prefix("0x").unwrap_or(addr).parse().map_err(ConversionError::FromHexError)?;
```

Comment (prestwich) [reply ethers-core/src/utils/mod.rs:441]: sorry this is a pain. if it gets too much let me know and i'll do the tidying

Comment (izayl) [reply ethers-core/src/utils/mod.rs]: @prestwich I've add this `FromHexError`

**Review by prestwich — APPROVED** [review]
Approving, pending CI run completion. Thanks for the contribution!

**Pull request merged**
